metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala

```diff
@@ -884,9 +884,9 @@ abstract class MetalsLspService(
   ): CompletableFuture[Unit] = {
     val path = params.getTextDocument.getUri.toAbsolutePath
     savedFiles.add(path)
-    // read file from disk, we only remove files from buffers on didClose.
-    val text = path.toInput.text
-    buffers.put(path, text)
+    if (!buffers.contains(path)) {
+      buffers.put(path, path.toInput.text)
+    }
```

Comment:
```suggestion
    buffers.putIfAbsent(path, path.toInput.text)
```
I think that `buffers` is based on a concurrent map (`TrieMap`), so maybe it would be safer to perform an atomic operation here, in case multiple requests are processed in parallel. :slightly_smiling_face: 

Also, a question to `didSave` and `didChange`: it looks like currently they always overwrite the in-memory `buffers` content for the given file, but in case of concurrent requests we could apply the older data (race condition)? Seems that the `DidCloseTextDocumentParams` should contain document version, could we use that to always safely keep only the newest one? Sorry if I have misunderstood how this all works. :slightly_smiling_face:

Reply:
Looks like I removed that part of code, but thanks for suggestion! We only now update on didOpen and didChange